### PR TITLE
MFA for App Access - `tsh proxy app` and Teleport Connect

### PR DESCRIPTION
**Pull request by Joerger**
Add support for longer-lived mfa-verified local proxy app certs for `tsh proxy app` and local app proxies started with Teleport Connect.

In a follow up PR I'll extend this logic for `tsh proxy aws|azure|gcp`.

Depends on https://github.com/gravitational/teleport/pull/40846.

### CODE REVIEW

File: tool/tsh/common/app.go [hunk 1]
```diff
@@ -52,26 +52,65 @@ func onAppLogin(cf *CLIConf) error {
...
+		return trace.Wrap(err)
+	}
+
 	app, err := getRegisteredApp(cf, tc)
 	if err != nil {
 		return trace.Wrap(err)
 	}
-	profile, err := tc.ProfileStatus()
+
+	routeToApp, err := getRouteToApp(cf, tc, profile, app)
```
Comment (Joerger) [inline tool/tsh/common/app.go:66]: Note to reviewers: the refactors in this file should have no functional implications - I just needed to reuse some of this logic. Some split-out methods were only needed for early iterations of this change, but it's easier (better?) not to revert these changes now.

File: lib/client/local_proxy_middleware.go [hunk 2]
```diff
@@ -29,60 +30,101 @@ import (
...
 // OnStart is a callback triggered when the local proxy starts.
-func (c *DBCertChecker) OnStart(ctx context.Context, lp *alpnproxy.LocalProxy) error {
+func (c *CertChecker) OnStart(ctx context.Context, lp *alpnproxy.LocalProxy) error {
 	return trace.Wrap(c.ensureValidCerts(ctx, lp))
 }
 
 // ensureValidCerts ensures that the local proxy is configured with valid certs.
-func (c *DBCertChecker) ensureValidCerts(ctx context.Context, lp *alpnproxy.LocalProxy) error {
-	if err := lp.CheckDBCerts(c.dbRoute); err != nil {
+func (c *CertChecker) ensureValidCerts(ctx context.Context, lp *alpnproxy.LocalProxy) error {
```
[the remark this answers is not in the document]
Comment (ravicious) [reply lib/client/local_proxy_middleware.go:83]: Its main purpose is to check cert expiration before the local proxy passes the connection to the target. In that case, `tsh proxy` and Connect will show you a prompt asking you to log in again.

In the case of database proxies, we also check if the subject in the cert matches the db route, but I don't remember the exact reason for that.

@greedy52 do you happen to remember why we're checking the subject?

https://github.com/gravitational/teleport/blob/97b16588f3c4fd100266c22414d22b7a0089727f/lib/srv/alpnproxy/local_proxy.go#L439-L456

File: lib/client/local_proxy_middleware.go [hunk 3]
```diff
@@ -29,60 +30,101 @@ import (
...
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// reduce per-handshake processing by setting the parsed leaf.
+	leaf, err := utils.TLSCertLeaf(cert)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	certTTL := leaf.NotAfter.Sub(c.clock.Now()).Round(time.Minute)
```
[the remark this answers is not in the document]
Comment (ravicious) [reply lib/client/local_proxy_middleware.go:100]: I think it's mostly so that the message says something like "valid for 60m" and not "valid for 59m59s".

File: tool/tsh/common/proxy.go [hunk 4]
```diff
@@ -317,16 +320,41 @@ func onProxyCommandApp(cf *CLIConf) error {
...
 		return trace.Wrap(err)
 	}
 
+	opts := []alpnproxy.LocalProxyConfigOpt{
+		alpnproxy.WithALPNProtocol(alpnProtocolForApp(app)),
+		alpnproxy.WithClusterCAsIfConnUpgrade(cf.Context, tc.RootClusterCACertPool),
+		alpnproxy.WithMiddleware(client.NewAppCertChecker(tc, routeToApp, nil)),
+	}
+
+	// If MFA is not required, try to load existing certs. If MFA is required,
```
Comment (ravicious) [inline tool/tsh/common/proxy.go:344]: I'm not an expert on this part of the codebase, but I have mixed feelings about keeping this logic in.

When would that be the case even? I guess someone would need to do `tsh app login` followed by `tsh proxy app`?

What I'm trying to say is that if we don't have tests for this scenario, keeping the logic for it around seems like a maintenance burden and possible source of bugs.

File: tool/tsh/common/app_test.go [hunk 5]
```diff
@@ -288,6 +288,16 @@ func TestAppCommands(t *testing.T) {
...
 								}()
 
+								if requireMFAType == types.RequireMFAType_SESSION {
+									// mfa verified proxy certs should not be saved to disk.
+									err = Run(context.Background(), []string{
+										"app",
+										"config",
+										app.name,
+									}, setHomePath(loginPath))
+									require.Error(t, err)
```
Comment (ravicious) [inline tool/tsh/common/app_test.go:298]: Should we put some additional expectations on this error? Like its type or message.

File: lib/auth/grpcserver.go [hunk 6]
```diff
@@ -2533,7 +2533,9 @@ func isLocalProxyCertReq(req *authpb.UserCertsRequest) bool {
 	return (req.Usage == authpb.UserCertsRequest_Database &&
 		req.RequesterName == authpb.UserCertsRequest_TSH_DB_LOCAL_PROXY_TUNNEL) ||
 		(req.Usage == authpb.UserCertsRequest_Kubernetes &&
-			req.RequesterName == authpb.UserCertsRequest_TSH_KUBE_LOCAL_PROXY)
+			req.RequesterName == authpb.UserCertsRequest_TSH_KUBE_LOCAL_PROXY) ||
+		(req.Usage == authpb.UserCertsRequest_App &&
+			req.RequesterName == authpb.UserCertsRequest_TSH_APP_LOCAL_PROXY)
```
Comment (ravicious) [inline lib/auth/grpcserver.go:2538]: I forgot to rebuild `teleport` and ran `tsh proxy app` against a cluster that didn't include this change.

`tsh proxy app` would ask for a tap and then exit silently. Re-running it in debug mode shows that an error was returned ("single-use certificates cannot be issued for app access").

Could you make sure that the error is surfaced to the user? This should help end users if starting a proxy fails for some other reason.

<details>
<summary>Error log from tsh app proxy -d</summary>

```
Tap any security key
Detected security key tap
2024-04-25T17:30:23+02:00 DEBU [CLIENT]    Issuing single-use certificate from unary GenerateUserCerts client/cluster_client.go:609
2024-04-25T17:30:23+02:00 ERRO [TSH]       Failed to start local proxy. error:[
ERROR REPORT:
Original Error: *interceptors.RemoteError single-use certificates cannot be issued for app access
Stack Trace:
	github.com/gravitational/teleport/api@v0.0.0/client/client.go:1102 github.com/gravitational/teleport/api/client.(*Client).GenerateUserCerts
	github.com/gravitational/teleport/lib/client/cluster_client.go:610 github.com/gravitational/teleport/lib/client.PerformMFACeremony
	github.com/gravitational/teleport/lib/client/cluster_client.go:373 github.com/gravitational/teleport/lib/client.(*ClusterClient).performMFACeremony
	github.com/gravitational/teleport/lib/client/cluster_client.go:486 github.com/gravitational/teleport/lib/client.(*ClusterClient).IssueUserCertsWithMFA
	github.com/gravitational/teleport/lib/client/api.go:1427 github.com/gravitational/teleport/lib/client.(*TeleportClient).IssueUserCertsWithMFA
	github.com/gravitational/teleport/lib/client/local_proxy_middleware.go:181 github.com/gravitational/teleport/lib/client.(*appCertReissuer).ReissueCert.func1
	github.com/gravitational/teleport/lib/client/api.go:592 github.com/gravitational/teleport/lib/client.RetryWithRelogin
	github.com/gravitational/teleport/lib/client/local_proxy_middleware.go:180 github.com/gravitational/teleport/lib/client.(*appCertReissuer).ReissueCert
	github.com/gravitational/teleport/lib/client/local_proxy_middleware.go:90 github.com/gravitational/teleport/lib/client.(*CertChecker).ensureValidCerts
	github.com/gravitational/teleport/lib/client/local_proxy_middleware.go:79 github.com/gravitational/teleport/lib/client.(*CertChecker).OnStart
	github.com/gravitational/teleport/lib/srv/alpnproxy/local_proxy.go:177 github.com/gravitational/teleport/lib/srv/alpnproxy.(*LocalProxy).start
	github.com/gravitational/teleport/lib/srv/alpnproxy/local_proxy.go:171 github.com/gravitational/teleport/lib/srv/alpnproxy.(*LocalProxy).Start
	github.com/gravitational/teleport/tool/tsh/common/proxy.go:387 github.com/gravitational/teleport/tool/tsh/common.onProxyCommandApp
	github.com/gravitational/teleport/tool/tsh/common/tsh.go:1454 github.com/gravitational/teleport/tool/tsh/common.Run
	github.com/gravitational/teleport/tool/tsh/common/tsh.go:594 github.com/gravitational/teleport/tool/tsh/common.Main
	github.com/gravitational/teleport/tool/tsh/main.go:26 main.main
	runtime/proc.go:271 runtime.main
	runtime/asm_arm64.s:1222 runtime.goexit
User Message: single-use certificates cannot be issued for app access] common/proxy.go:388
```
</details>

File: lib/client/local_proxy_middleware.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (greedy52) [reply lib/client/local_proxy_middleware.go:83]: Backend gets db route info from the cert so I think this is just double-checking that the cert has the same correct route info as the user requsted.

If local proxy creation and cert renewal logic are implemented correctly for Connect, this check is probably a good to have but unnecessary. Legacy `tsh` allows users to specify a cert file path as a parameter. Also when not doing per-session MFA, `tsh` loads certs from the file system. So i guess it's good to check in those cases.

File: lib/teleterm/clusters/cluster_apps.go
Comment (ravicious) [inline lib/teleterm/clusters/cluster_apps.go:1]: Could you add integration tests for app gateways with MFA?

https://github.com/gravitational/teleport/blob/bbcc5e2b8a675f37d30b1e80155a97416736402f/integration/proxy/teleterm_test.go#L601-L619

The implementation can be based on the tests for kube gateways. I'm just not sure how to pluck `auth.Server` out of `appaccess.Pack`.

Ping me if you run into problems. The MFA tests in teleterm are a huge mess because of how many things depend on each other.

https://github.com/gravitational/teleport/blob/bbcc5e2b8a675f37d30b1e80155a97416736402f/integration/proxy/teleterm_test.go#L413-L446

**Review by ravicious — COMMENTED** [review]
I tested the proxy in both tsh and Connect and both seem to work well.

File: lib/teleterm/clusters/cluster_apps.go [hunk 7]
```diff
@@ -181,27 +182,18 @@ func (c *Cluster) reissueAppCert(ctx context.Context, clusterClient *client.Clus
 		return tls.Certificate{}, trace.Wrap(err)
 	}
 
-	err = clusterClient.ReissueUserCerts(ctx, client.CertCacheKeep, client.ReissueParams{
+	key, _, err := clusterClient.IssueUserCertsWithMFA(ctx, client.ReissueParams{
```
[the remark this answers is not in the document]
Comment (ravicious) [reply lib/teleterm/clusters/cluster_apps.go:185]: As I mentioned during the call, when per-session MFA is needed, certs for local proxies are stored in memory instead of on disk. Check [RFD 90](https://github.com/gravitational/teleport/blob/master/rfd/0090-db-mfa-sessions.md) for more details.

File: lib/srv/alpnproxy/local_proxy.go [hunk 8]
```diff
@@ -378,6 +378,27 @@ func (l *LocalProxy) getCerts() []tls.Certificate {
...
 
+// CheckCertExpiry checks the proxy certificates for expiration and runs given checking function.
+func (l *LocalProxy) CheckCert(checkCert func(cert *x509.Certificate) error) error {
+	l.cfg.Log.Debug("checking local proxy certs")
+	l.certsMu.RLock()
+	defer l.certsMu.RUnlock()
+	if len(l.cfg.Certs) == 0 {
+		return trace.NotFound("local proxy has no TLS certificates configured")
+	}
+	cert, err := utils.TLSCertLeaf(l.cfg.Certs[0])
```
[the remark this answers is not in the document]
Comment (Joerger) [reply lib/srv/alpnproxy/local_proxy.go:389]: Looks like you are right, we are never using more than one cert. Adjusted it to be singular across the board.

FWIW I think the original reason was to make a one-to-one translation to `tls.Config.Certificates`, but I think the verbosity is not worth it since it's not actually used pluraly.

File: tool/tsh/common/proxy.go (hunk 4 above)
Comment (Joerger) [reply tool/tsh/common/proxy.go:344]: This was meant to maintain feature parity with `tsh proxy kube|db` but you're right that it's not necessary. It actually makes more sense to me to have `tsh proxy app` create a new session separate from any ongoing `tsh app login` credentials, so the proxy session can be its own. Thanks for raising this.

File: lib/teleterm/clusters/cluster_apps.go
Comment (Joerger) [reply lib/teleterm/clusters/cluster_apps.go:1]: Yeah I can see what you mean about the mess, the app access integration `Pack` is very unapproachable, but hard to refactor at this point. I made it work but didn't bother to make things any cleaner...

**Review by ravicious — APPROVED** [review]

**Review by bl-nero — APPROVED** [review]

File: lib/srv/alpnproxy/local_proxy.go [hunk 9]
```diff
@@ -434,20 +461,20 @@ func CheckCertSubject(cert *x509.Certificate, dbRoute tlsca.RouteToDatabase) err
...
-// getCertsForConn determines if certificates should be used when dialing
+// getCerForConn determines if certificates should be used when dialing
 // upstream to proxy a new downstream connection.
-// After calling getCertsForConn function, the returned
+// After calling getCerForConn function, the returned
 // net.Conn should be used for further operation.
-func (l *LocalProxy) getCertsForConn(ctx context.Context, downstreamConn net.Conn) ([]tls.Certificate, net.Conn, error) {
-	if !l.cfg.CheckCertsNeeded {
-		return l.getCerts(), downstreamConn, nil
+func (l *LocalProxy) getCerForConn(downstreamConn net.Conn) (tls.Certificate, net.Conn, error) {
```
Comment (nklaassen) [inline lib/srv/alpnproxy/local_proxy.go:475]:
```suggestion
func (l *LocalProxy) getCertForConn(downstreamConn net.Conn) (tls.Certificate, net.Conn, error) {
```

File: lib/client/local_proxy_middleware.go [hunk 10]
```diff
@@ -29,103 +30,194 @@ import (
...
+
+func (c *AppCertIssuer) CheckCert(cert *x509.Certificate) error {
+	// appCertIssuer does not perform any additional certificate checks.
+	return nil
+}
+
+func (c *AppCertIssuer) IssueCert(ctx context.Context) (tls.Certificate, error) {
+	var accessRequests []string
+	if profile, err := c.Client.ProfileStatus(); err != nil {
+		log.WithError(err).Warn("unable to load profile, requesting database certs without access requests")
```
Comment (nklaassen) [inline lib/client/local_proxy_middleware.go:183]:
```suggestion
		log.WithError(err).Warn("unable to load profile, requesting app certs without access requests")
```

File: lib/client/local_proxy_middleware.go [hunk 11]
```diff
@@ -29,103 +30,194 @@ import (
...
+
+		// TODO (Joerger): DELETE IN v17.0.0
+		clusterClient, err := c.Client.ConnectToCluster(ctx)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		rootClient, err := clusterClient.ConnectToRootCluster(ctx)
+		if err != nil {
+			return trace.Wrap(err)
+		}
```
Comment (nklaassen) [inline lib/client/local_proxy_middleware.go:205]: does this dial the root cluster? do you think it would be possible/worth avoiding it, if clusterClient is the root cluster, or the session isn't necessary? could be a followup PR, i know this is already used elsewhere

**Review by nklaassen — APPROVED** [review]

File: lib/client/local_proxy_middleware.go (hunk 11 above)
Comment (Joerger) [reply lib/client/local_proxy_middleware.go:205]: If `clusterClient` is root, it just returns the auth client already dialed:

```go
// ConnectToRootCluster connects to the auth server of the root cluster
// via proxy. It returns connected and authenticated auth server client.
func (c *ClusterClient) ConnectToRootCluster(ctx context.Context) (auth.ClientI, error) {
	root, err := c.ConnectToCluster(ctx, c.root)
	return root, trace.Wrap(err)
}

// ConnectToCluster connects to the auth server of the given cluster via proxy. It returns connected and authenticated auth server client
func (c *ClusterClient) ConnectToCluster(ctx context.Context, clusterName string) (auth.ClientI, error) {
	if c.cluster == clusterName {
		return c.CurrentCluster(), nil
	}
...
```

**Comment by Joerger** [comment]
@zmb3 @r0mant Can I get an `/excludeflake *`

**Comment by Joerger** [comment]
@zmb3 Can I please get an `/excludeflake *`
